incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,90 @@
+package stats
+
+import (
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+func (c *counter) Incr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, 1)
+	*c = counter(i)
+}
+
+func (c *counter) Decr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, -1)
+	*c = counter(i)
+}
+
+type stats struct {
+	// HcTot is the total number of HierarchyConfig reconciliations.
+	HcTot counter
+
+	// HcCur is the currently undergoing number of HierarchyConfig reconciliations.
+	HcCur counter
+
+	// ObTot is the total number of object reconciliations.
+	ObTot counter
+
+	// ObCur is the currently undergoing number of object reconciliations.
+	ObCur counter
+
+	// HcRead is the total number of HierarchyConfig reads.
+	HcRead counter
+
+	// HcWrite is the total number of HierarchyConfig writes.
+	HcWrite counter
+
+	// ObRead is the total number of object reads.
+	ObRead counter
+
+	// ObWrite is the total number of object writes.
+	ObWrite counter
+}
+
+var Stats stats
+
+// LogActivity generates logs for performance testing.
+func LogActivity() {
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, cur counter = 0, 0
+	working := false
+	actN := 1
+	go func() {
+		// run forever
+		for {
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			if Stats.HcTot+Stats.ObTot != total || cur != 0 {
```

Comment:
I find this logic a bit hard to read. Also, why are we comparing the _current_ values for the totals, but the _last_ values for `cur`?

Maybe rename `total` to `lastTotal`, and then start this block as follows:

```
total = HcTot + ObjTot
cur = HcCur + ObjCur
if total != lastTotal || cur > 0 {
  ...
}
lastTotal = total
```

Reply:
I used the `lastCur != 0` to still output the log for what happened in the past 0.5s. Otherwise, the last active log will be missing.